### PR TITLE
fix: badge count WPB-453

### PR DESCRIPTION
**Pull request by KaterinaWire**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-453" title="WPB-453" target="_blank">BugWPB-453</a>  [iOS] Badge count falsely shows unread messages
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

The app icon badge count shows the wrong number of conversations that have unread messages.

### Causes (Optional)

**Expected flow:**
1. Push notification increases the badge count. _(Notification extension)_
2. The app in the foreground handles events and reduces the badge count. _(Main app)_

For some reason (and for some periods of time) the Notifications extension saves its changes while the app is in the foreground and overrides the correct number of unread messages.

### Solutions

Recalculate unread messages when the app enters background.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-notification-engine/Sources/NotificationSession.swift [hunk 1]
```diff
@@ -362,6 +362,7 @@ extension NotificationSession: PushNotificationStrategyDelegate {
                 WireLogger.notifications.info("ignoring event")
             }
         }
+        context.saveOrRollback()
```
Comment (johnxnguyen) [inline wire-ios-notification-engine/Sources/NotificationSession.swift:365]: What was this solving?

Comment (KaterinaWire) [reply wire-ios-notification-engine/Sources/NotificationSession.swift:365]: Save changes and increase the badge count. If you check the develop branch, the badge count will always be 1.

**Review by netbe — COMMENTED** [review]
@KaterinaWire is there a way to add a regression test on it ?

Note once merged this needs to be backport to 3.110 branch for a hotfix release

**Comment by KaterinaWire** [comment]
> @KaterinaWire is there a way to add a regression test on it ?
> 
> Note once merged this needs to be backport to 3.110 branch for a hotfix release

@netbe  What do you mean by regression test? This issue is not possible to reproduce every time with certain steps.

**Review by netbe — APPROVED** [review]

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**